### PR TITLE
Perform resize of mounted volume if necessary

### PR DESCRIPTION
**Pull request by gnufied**
Under certain conditions - we must perform resize of volume even when it is mounted. This enables us to get around problem of resizing volumes used by deployments etc.

```release-note
Allow expanding mounted volumes
```

### CODE REVIEW

File: pkg/util/resizefs/resizefs_linux.go [hunk 1]
```diff
@@ -61,13 +61,19 @@ func (resizefs *ResizeFs) Resize(devicePath string) (bool, error) {
...
 	if err != nil {
-		deviceOpenErr := fmt.Errorf("error verifying if device %s is open: %v", devicePath, err)
+		deviceOpenErr := fmt.Errorf("ResizeFS.Resize - error verifying if device %s is open: %v", devicePath, err)
 		return false, deviceOpenErr
 	}
 
 	if deviceOpened {
-		deviceAlreadyOpenErr := fmt.Errorf("the device %s is already in use", devicePath)
-		return false, deviceAlreadyOpenErr
+		glog.Warningf("ResizeFS.Resize - Expanding mounted volume %s", devicePath)
```
Comment (jsafrane) [inline pkg/util/resizefs/resizefs_linux.go]: Why it's `Warning`? It should be normal operation.

File: test/e2e/storage/mounted_volume_resize.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+		pvcClaims := []*v1.PersistentVolumeClaim{pvc}
+		pvs, err := framework.WaitForPVClaimBoundPhase(c, pvcClaims, framework.ClaimProvisionTimeout)
+		Expect(err).NotTo(HaveOccurred(), "Failed waiting for PVC to be bound %v", err)
+		Expect(len(pvs)).To(Equal(1))
+
+		By("Creating a deployment with the provisioned volume")
+		deployment, err := framework.CreateDeployment(c, int32(1), map[string]string{"test": "app"}, nodeKeyValueLabel, ns, pvcClaims, "")
+		defer c.ExtensionsV1beta1().Deployments(ns).Delete(deployment.Name, &metav1.DeleteOptions{})
+
+		By("Expanding current pvc")
```
Comment (jsafrane) [inline test/e2e/storage/mounted_volume_resize.go]: Should you wait for a pod to be Running before resize? Expand controller may be faster than the pod.

**Review by jsafrane — COMMENTED** [review]
I see what you're trying to do, but IMO we should do resize even if a pod is `Running` - both ext4 and xfs allow online resize and ext3 allows it in most cases too. Why should user restart pods?

[note that ext3 without resize_inode option will fail also in this PR, the volume must be fully unmounted first]

**Comment by gnufied** [comment]
@jsafrane proposal for implementing online resize is here - https://github.com/kubernetes/community/pull/1535 , so we will get there. But there are design problems to be solved to get there.

File: test/e2e/storage/mounted_volume_resize.go (hunk 2 above)
Comment (gnufied) [reply test/e2e/storage/mounted_volume_resize.go]: The `CreateDeoployment` function waits for a pod to be running.

**Comment by gnufied** [comment]
@jsafrane also -  I checked RHEL, Fedora 26 and 27, Ubuntu, Debian and Google COS - they all have `resize_inode` option enabled in `mke2fs.conf`. For now - may be we can document this (until we get out of alpha). Not all PVCs are resizable anyways, only those that are explicitly enabled by kube admin. I am not sure what else we can do in this release.

**Comment by jsafrane** [comment]
> I am not sure what else we can do in this release

Just throw a sensible event on the PVC that online resize failed with error from stderr. Let's hope it's enough to the users to either try offline resize or nag system admin to resize the volume manually (e.g. because is has errors).

File: test/e2e/storage/mounted_volume_resize.go (hunk 2 above)
Comment (jsafrane) [reply test/e2e/storage/mounted_volume_resize.go]: ack

File: pkg/util/resizefs/resizefs_linux.go (hunk 1 above)
Comment (jsafrane) [reply pkg/util/resizefs/resizefs_linux.go]: can you please fix the level? I'm inclined to approve this PR then.

Comment (gnufied) [reply pkg/util/resizefs/resizefs_linux.go]: fixed

**Comment by gnufied** [comment]
The failure looks like unrelated flake or breakage - I have filed https://github.com/kubernetes/kubernetes/issues/58881 . I did not see an existing ticket for the failure.

**Comment by gnufied** [comment]
@jsafrane PTAL, changed log level of resize message.

File: pkg/util/resizefs/resizefs_linux.go [hunk 3]
```diff
@@ -84,7 +90,7 @@ func (resizefs *ResizeFs) Resize(devicePath string) (bool, error) {
 		}
 		return resizefs.xfsResize(devicePath)
```
Comment (mlmhl) [inline pkg/util/resizefs/resizefs_linux.go]: I think the resize operation maybe can't work correctly for `xfs` file system: `resizeFileSystem` method is performed between attaching device and mounting device, it means the resize operation maybe performed before the device mounted. But `xfs_growfs` requires the device has been mounted. So if we try to resize a volume with `xfs` file system, we may receive an error event like this:

```
MountVolume.resizeFileSystem failed for volume "pvc-60265674-0406-11e8-9c91-0800273c9701" : resize of device /dev/rbd0 failed: exit status 1. xfs_growfs output: xfs_growfs: /dev/rbd0 is not a mounted XFS filesystem
```

Comment (gnufied) [reply pkg/util/resizefs/resizefs_linux.go]: Thank you for catching this. It is fixed now.

**Comment by gnufied** [comment]
@jsafrane PTAL.

**Pull request merged**